bootstore/src/node.rs

```diff
@@ -63,54 +69,241 @@ impl Node {
             NodeOp::KeySharePrepare {
                 rack_uuid,
                 epoch,
                 share_distribution,
             } => self.handle_key_share_prepare(
-                rack_uuid,
+                &rack_uuid,
                 epoch,
                 share_distribution,
             ),
-            NodeOp::KeyShareCommit { rack_uuid, epoch } => {
-                self.handle_key_share_commit(rack_uuid, epoch)
-            }
+            NodeOp::KeyShareCommit {
+                rack_uuid,
+                epoch,
+                prepare_share_distribution_digest,
+            } => self.handle_key_share_commit(
+                &rack_uuid,
+                epoch,
+                prepare_share_distribution_digest,
+            ),
         };
 
         NodeResponse { version: req.version, id: req.id, result }
     }
 
+    // Handle `GetShare` messages from another node
     fn handle_get_share(
         &mut self,
-        _epoch: i32,
+        epoch: i32,
     ) -> Result<NodeOpResult, NodeError> {
-        unimplemented!();
+        let share = self.db.get_committed_share(&mut self.conn, epoch)?;
+        Ok(NodeOpResult::Share { epoch, share: share.0.share })
     }
 
+    // Handle `Initialize` messages from the coordinator
     fn handle_initialize(
         &mut self,
-        _rack_uuid: Uuid,
-        _share_distribution: SerializableShareDistribution,
+        rack_uuid: &Uuid,
+        share_distribution: SerializableShareDistribution,
     ) -> Result<NodeOpResult, NodeError> {
-        unimplemented!();
+        self.db.initialize(&mut self.conn, rack_uuid, share_distribution)?;
+        Ok(NodeOpResult::CoordinatorAck)
     }
 
+    // Handle `KeySharePrepare` messages from the coordinator
     fn handle_key_share_prepare(
         &mut self,
-        _rack_uuid: Uuid,
-        _epoch: i32,
-        _share_distribution: SerializableShareDistribution,
+        rack_uuid: &Uuid,
+        epoch: i32,
+        share_distribution: SerializableShareDistribution,
     ) -> Result<NodeOpResult, NodeError> {
-        unimplemented!();
+        if epoch == 0 {
+            return Err(NodeError::KeySharePrepareForEpoch0);
+        }
+
+        self.db.prepare_share(
+            &mut self.conn,
+            rack_uuid,
+            epoch,
+            share_distribution,
+        )?;
+
+        Ok(NodeOpResult::CoordinatorAck)
     }
 
+    // Handle `KeyShareCommit` messages from the coordinator
     fn handle_key_share_commit(
         &mut self,
-        _rack_uuid: Uuid,
-        _epoch: i32,
+        rack_uuid: &Uuid,
+        epoch: i32,
+        prepare_shared_distribution_digest: Sha3_256Digest,
     ) -> Result<NodeOpResult, NodeError> {
-        unimplemented!();
+        self.db.commit_share(
+            &mut self.conn,
+            rack_uuid,
+            epoch,
+            prepare_shared_distribution_digest,
+        )?;
+
+        Ok(NodeOpResult::CoordinatorAck)
+    }
+}
+
+#[cfg(test)]
+mod tests {
```

Comment:
I found these tests to be eminently readable, and they helped me better understand the implementation :+1:

Reply:
Thanks! I almost recommended reading them first. Maybe I should do that next time.